core/embed/trezorhal/rgb_led.c

```diff
@@ -0,0 +1,130 @@
+
+#include "common.h"
+
+#include STM32_HAL_H
+
+#define RESET_DATA_LEN 18
+#define DATA_LEN 25
+#define TIMER_PERIOD 832  // cca 200 KHz @ 180MHz
+#define BIT_0_LEN 52      // 312ns
+#define BIT_1_LEN 125     // 750ns
+
+#if defined BOARDLOADER
+#error Not implemented for boardloader!
+#endif
+
+#if defined BOOTLOADER
+__attribute__((section(".buf")))
+#endif
+uint32_t rgb_led_data[RESET_DATA_LEN + DATA_LEN] = {0};
+
+static void rgb_led_set(uint32_t* start, uint8_t color) {
```

Comment:
What kind of protocol is this?

Maybe we can add some comments explaining what this function does?

Are we setting PWM delays with DMA? Why do constants `52` and `125` have ns equivalent? 

I have trouble understanding what exactly is going on. So maybe having a short explanation about the high-level architecture would be nice.

Reply:
In the datasheet they call it "unipolar zeroing code" and basically they prescribe short pulse (200-400ns) for 0 and long pulse (580ns-1us) for 1. Reset is required before sending data: >80us without pulse. Then the data is sent -> RGB encoded as long and short pulses, so 24 bits (24 values of compare levels for PWM), and final value is zero to stop pulsing.  

PWM compare levels are set via DMA to avoid need for interrupt and manual change after every bit is sent.

I will add some explanation to the source also.

Reply:
done 4f6c9f55d9168d4f72e7cfac7f302776fa4b4f3c